File: 022/cpp/main.cpp
class Solution {
public:
    vector<string> f(int left, int right) {
        if (left == 0) {
            return {std::string(right, ')')};
        }
        
        std::vector<string> r;

        std::vector<string> r0 = f(left - 1, right + 1);
        for (auto &s: r0) {
            r.push_back("(" + s);
        }

        if (right > 0) {
            std::vector<string> r1 = f(left, right - 1);
            for (auto &s: r1) {
                r.push_back(")" + s);
            }
        }

        return r;
    }

    vector<string> generateParenthesis(int n) {
        return f(n, 0);
    }
};

File: 022/rust/main.rs
impl Solution {
    pub fn f(left: i32, right: i32) -> Vec<String> {
        if left == 0 {
            return vec![")".repeat(right as usize)];
        }
        
        let mut r = Vec::new();
        
        let r0 = Solution::f(left - 1, right + 1);
        for s in r0 {
            r.push(format!("({s}"));
        }

        if (right > 0) {
            let r0 = Solution::f(left, right - 1);
            for s in r0 {
                r.push(format!("){s}"));
            }
        }

        return r;
    }

    pub fn generate_parenthesis(n: i32) -> Vec<String> {
        return Solution::f(n, 0);
    }
}
